src/backend/distributed/commands/utility_hook.c

```diff
@@ -517,13 +517,42 @@ multi_ProcessUtility(PlannedStmt *pstmt,
 	{
 		PostStandardProcessUtility(parsetree);
 
+		/*
+		 * We call MarkInvalidateForeignKeyGraph in preprocess without knowing
+		 * that if command would fail or not. However, we don't want to invalidate
+		 * foreign key graph for failed DDL commands, so here we don't call
+		 * InvalidateForeignKeyGraphForDDL.
+		 * On the other hand, we should still reset shouldInvalidateForeignKeyGraph
+		 * flag as next DDL command still relies on this flag to invalidate foreign
+		 * key graph.
+		 */
+		shouldInvalidateForeignKeyGraph = false;
+
 		PG_RE_THROW();
 	}
 	PG_END_TRY();
 
 	/*
 	 * Post process for ddl statements
 	 */
+
+	if (IsA(parsetree, CreateStmt))
+	{
+		CreateStmt *createStatement = (CreateStmt *) parsetree;
+
+		PostprocessCreateTableStmt(createStatement, queryString);
+	}
+
+	/*
+	 * Re-forming the foreign key graph relies on the command being executed
+	 * on the local table first. However, in order to decide whether the
+	 * command leads to an invalidation, we need to check before the command
+	 * is being executed since we read pg_constraint table. Thus, we maintain a
+	 * local flag and do the invalidation after multi_ProcessUtility,
+	 * before ExecuteDistributedDDLJob().
+	 */
+	InvalidateForeignKeyGraphForDDL();
```

Comment:
isn't it safer to call this at the end of the function?

Reply:
then moving this block right before `ExecuteDistributedDDLJob` part, as comment implies